### ISSUE TITLE
x509 pem parsing fails on CT Precertificate OID (1.3.6.1.4.1.11129.2.4.3)

### ISSUE THREAD

**Issue by asayler** [pyca/cryptography#3325]
When trying to import pem data for an x509 CT precert, cryptography.x509 raises the following error:

```
UnsupportedExtension: Critical extension <ObjectIdentifier(oid=1.3.6.1.4.1.11129.2.4.3, name=Unknown OID)> is not currently supported
```

Presumably, this is due to a lack of support for the Certficate Transperency precertificate OID in cryptography: http://www.alvestrand.no/objectid/submissions/1.3.6.1.4.1.11129.2.4.3.html

It would be useful to add support for this OID so that cryptography can be used to parse CT precert data in addition to normal certs.

**Comment by alex** [comment]
So is the intent of the poison extension that it prevents ever treating a pre-cert as a cert, or just that it prevents validators from considering precerts as certs?

Basically: do we just add poison processing to load_cert, or do we need a load_precert?

**Comment by asayler** [comment]
That's a good question...

The [RFC](https://tools.ietf.org/html/rfc6962) states that:

> The Precertificate is constructed from the certificate to be issued by adding a special critical poison extension (OID 1.3.6.1.4.1.11129.2.4.3, whose extnValue OCTET STRING contains ASN.1 NULL data (0x05 0x00)) to the end-entity TBSCertificate (this extension is to ensure that the Precertificate cannot be validated by a standard X.509v3 client)...

So if we want to comply with that, it would seem wise to not just silently accepting precerts. Maybe load_cert could take an `allow_precert=False` flag? Or use a separate function as you suggest.

But maybe someone with more knowledge of certificate transparency in general could comment on the wisdom or danger of processing precerts and real certs with teh same function.

**Comment by alex** [comment]
From Ryan Sleevi:

> @asayler @alex_gaynor Related to the CT discussion in #urandom - parsers shouldn't reject unknown critical extensions, validators should
> 5280 covers the "what to do with unknown"
> so you add the extension to your parser, but update your validator to reject any •unhandled• critical extensions
> That's the layering design at least

**Comment by sleevi** [comment]
Right, because X.509v3 is based on ASN.1/DER, the goal is to ensure a parser can be context-aware and the protocol extended, without having to update parsers (hence, why all the tag-length-value nonsense). Similarly, X.509v3 extensions are meant to be extensible - you more or less ignore anything you don't know how to parse (hence the OCTETSTRING wrapping)

All the 'meat' of the logic comes from RFC5280 - which defines which extensions you MUST recognize (at least, for the 5280 profile), and what to do with unhandled critical extensions (whether or not you recognize them, it's about the ones you don't mark as processed). You want to reject any critical extension whose syntax regarding validation you don't understand.

So for CT, you could just ignore the unknown extension - and leave it to applications/downstream consumers to include CT parsing logic. You could recognize it yourself, and then represent it however you want (whether simply saying "I know the symbolic name for it" or doing a more structured parse of it and exposing that information as part of your API).

However, when it comes to validating, any 2459/3280/5280 'compatible' validator should say "Hey, there's a critical extension, I don't know how it fits, I'm out". Any compatible validator that knows about the CT precert knows it's a "Don't trust this cert, I'm out" flag. So both result in the same observable effect.

TL;DR: I suggest being less strict on parsing, but make sure you've got the validation right. And if you're just a parser for downstream validators, well, they should have read/implemented the RFC, that's what it's there for :)
